components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/action/execution/PreIssueAccessTokenRequestBuilder.java

```diff
@@ -112,12 +120,57 @@ private Event getEvent(OAuthTokenReqMessageContext tokenMessageContext, Map<Stri
             eventBuilder.userStore(new UserStore(authorizedUser.getUserStoreDomain()));
         }
 
-        eventBuilder.accessToken(getAccessToken(tokenMessageContext, claimsToAdd));
+        // Retrieving app information here to reduce the perf overhead.
+        OAuthAppDO oAuthAppDO;
+        try {
+            oAuthAppDO = getAppInformation(tokenMessageContext);
+        } catch (IdentityOAuth2Exception | InvalidOAuthClientException e) {
+            throw new ActionExecutionRequestBuilderException(
+                    "Failed to generate pre issue access token action request for application: " +
+                            tokenMessageContext.getOauth2AccessTokenReqDTO().getClientId() + " grant type: " +
+                            tokenMessageContext.getOauth2AccessTokenReqDTO().getGrantType(), e);
+        }
+
+        eventBuilder.accessToken(getAccessToken(tokenMessageContext, oAuthAppDO, claimsToAdd));
+
+        if (isRefreshTokenAllowed(oAuthAppDO)) {
+            eventBuilder.refreshToken(getRefreshToken(oAuthAppDO.getRefreshTokenExpiryTime(),
+                    tokenMessageContext.getRefreshTokenValidityPeriodInMillis()));
+        }
         eventBuilder.request(getRequest(tokenReqDTO));
 
         return eventBuilder.build();
     }
 
+    private boolean isRefreshTokenAllowed(OAuthAppDO oAuthAppDO) {
+
+        if (OAuthServerConfiguration.getInstance().getSupportedGrantTypes().containsKey(
+                GrantType.REFRESH_TOKEN.toString()) && oAuthAppDO != null) {
+
+            String grantTypes = oAuthAppDO.getGrantTypes();
+            if (StringUtils.isNotEmpty(grantTypes)) {
+                List<String> supportedGrantTypes = Arrays.asList(grantTypes.split(" "));
+                return supportedGrantTypes.contains(OAuthConstants.GrantTypes.REFRESH_TOKEN);
+            }
+        }
+        return false;
+    }
+
+    private RefreshToken getRefreshToken(long refreshTokenExpiryInSecondsFromApp,
+                                         long refreshTokenValidityPeriodInMillisFromContext) {
+
+        long refreshTokenValidityPeriod = -1;
+        if (refreshTokenValidityPeriodInMillisFromContext > 0) {
+            // To honor the existing custom refresh grant handling.
+            refreshTokenValidityPeriod = TimeUnit.MICROSECONDS.toSeconds(refreshTokenValidityPeriodInMillisFromContext);
+        } else if (refreshTokenExpiryInSecondsFromApp > 0) {
+            refreshTokenValidityPeriod = refreshTokenExpiryInSecondsFromApp;
+        }
```

Comment:
This logic needs more explanation
Looks like you have given precendence to expiry time in message context better mention why

Reply:
Fixed 49caebcb5cca897a855b8ca05cb5db3e23952e9d